notification_service/main.py

```diff
@@ -0,0 +1,378 @@
+                    },
+                },
+            ]
+        }
+
+        if queue_name == "project_create":
+            message_format["blocks"][0]["text"]["text"] = "Project Created on GitLab"
+        elif queue_name == "project_rename":
+            message_format["blocks"][0]["text"]["text"] = "Project Renamed on GitLab"
+        elif queue_name == "project_transfer":
+            message_format["blocks"][0]["text"][
+                "text"
+            ] = "Project Ownership Transferred on GitLab"
+
+    return message_format
+
+
+# SlackNotifier class, inherits from EventProcessor.
+# Used to retrieve events from Redis and send them to Slack.
+class SlackNotifier(EventProcessor):
+    def __init__(self, slack_webhook_url, queue_prefix, events, redis_conn=None):
+        super().__init__(queue_prefix, events, redis_conn)
+        self.slack_webhook_url = slack_webhook_url
+
+        prometheus_multiproc_dir = "prometheus_multiproc_dir"
+
+        os.makedirs(prometheus_multiproc_dir, exist_ok=True)
+
+        registry = CollectorRegistry()
+        multiprocess.MultiProcessCollector(registry)
+
+        self.notification_counter = Counter(
+            "notification_service_notifications_total",
+            "The number of times a notification has been sent",
+            ["notification_endpoint"],
+        )
+
+        self.notification_failures_counter = Counter(
+            "notification_service_notification_failures_total",
+            "The number of times a notification has failed to be sent",
+            ["notification_endpoint"],
+        )
+
+        self.notification_latency_histogram = Histogram(
+            "notification_service_notification_latency_seconds",
+            "Time taken to send a notification",
+        )
+
+        self.queue_size_gauge = Gauge(
+            "notification_service_queue_size", "Number of events waiting in the queue"
+        )
+
+    def process_event(self, queue_name, data):
+        formatted_message = format_message(queue_name, data)
+
+        with self.notification_latency_histogram.time():
+            response = requests.post(self.slack_webhook_url, json=formatted_message)
+
+        if response.status_code != 200:
+            logging.debug(
+                f"Failed to send message to Slack: queue name: {queue_name}, data: {formatted_message}. "
+                f"Response code: {response.status_code} message: {response.content}"
+            )
+            self.notification_failures_counter.labels(self.slack_webhook_url).inc()
+        else:
+            logging.debug("Successfully sent message to Slack")
+            self.notification_counter.labels(self.slack_webhook_url).inc()
+
+
+def main(
+    slack_webhook_url="",
+    redis_conn=None,
+    testing=False,
+):
+    notifier = SlackNotifier(
+        slack_webhook_url, "classification", event_types, redis_conn=redis_conn
+    )
+
+    while True:
+        notifier.retrieve_event()
```

Comment:
You have no concurrency here. If the events are io bound then simply by using asyncio and the redis pub/sub library you can have concurent execution of io tasks rather than always handling only one event at a time.